### ISSUE TITLE
Early exiting callbacks

### ISSUE THREAD

**Issue by joehuchette** [JuliaOpt/MathProgBase.jl#59]
There are three approaches I can think of:

1. Check the return value and exit if ``== :Exit``
2. Add a new interface function (``abortcallback``?) that exits
3. Do nothing and just use ``error``s in Julia

I think I prefer 2), since 1) can easily lead to type instability. Thoughts? @mlubin @IainNZ

**Comment by mlubin** [comment]
What about having a specific Exception type that you can throw to exit?

**Comment by joehuchette** [comment]
Seems reasonable, assuming the try/catch blocks don't mean a performance hit.

**Comment by joehuchette** [comment]
Try/catch adds about 5% to the runtime of Iains benchmark:

Try/catch 1
```
minimum(times) = 10.097314416
median(times) = 10.5421976445
mean(times) = 11.001705035250001
```
Try/catch 2
```
minimum(times) = 10.024404999
median(times) = 10.415319063
mean(times) = 11.15705522385
```
Master 1
```
minimum(times) = 10.003989751
median(times) = 10.0912865695
mean(times) = 10.24103035735
```
Master 2
```
minimum(times) = 9.978583618
median(times) = 10.046673308999999
mean(times) = 10.146638271699999
```

**Comment by mlubin** [comment]
How did you implement it?

**Comment by joehuchette** [comment]
```jl
    function cutcallback(d::MathProgBase.MathProgCallbackData)
        state = MathProgBase.cbgetstate(d)
        @assert state == :MIPSol || state == :MIPNode
        if state == :MIPSol  # This shouldn't happen right?
            println("Is this ever called?")
            MathProgBase.cbgetmipsolution(d,m.colVal)
        else
            MathProgBase.cbgetlpsolution(d,m.colVal)
        end
        for cb in cbs
            try
                cb.f(d)
            catch y
                if isa(y, CallbackAbort)
                    return :Exit
                else
                    rethrow(y)
                end
            end
        end
        :Continue
    end
```

**Comment by mlubin** [comment]
Put it outside the loop?

**Comment by joehuchette** [comment]
Same.
```
minimum(times) = 10.414050885
median(times) = 10.664140355499999
mean(times) = 10.98764133475
```
with 
```jl
function lazycallback(d::MathProgBase.MathProgCallbackData)
    state = MathProgBase.cbgetstate(d)
    @assert state == :MIPSol || state == :MIPNode
    if state == :MIPSol
        MathProgBase.cbgetmipsolution(d,m.colVal)
    elseif anyfrac
        MathProgBase.cbgetlpsolution(d,m.colVal)
    else
        return :Continue
    end
    try
        for cb in cbs
            if state == :MIPSol || cb.fractional
                cb.f(d)
            end
        end
    catch y
        if isa(y, CallbackAbort)
            return :Exit
        else
            rethrow(y)
        end
    end
    :Continue
end
```

**Comment by mlubin** [comment]
I think that's fine. 5% is an upper bound on the overhead for the case when the callbacks do nothing.

**Comment by IainNZ** [comment]
Seems fine

**Comment by joehuchette** [comment]
I'll open a PR, but I'm not sure this is the right fix. How often is this actually used? Why is taking a 5-10% performance hit across the board for it?

**Comment by mlubin** [comment]
Here's an idea to lower the overhead: make ``lazycallback`` etc top-level functions so that they're properly compiled and type inferred and call them using anonymous closures.

**Comment by joehuchette** [comment]
Neat trick, looks like it worked:
```
minimum(times) = 10.081697514
median(times) = 10.139390735
mean(times) = 10.15984297695
```

**Comment by mlubin** [comment]
So we've settled on the ``:Exit`` convention, we should add this to the docs.

**Comment by mlubin** [comment]
Resolved with f8859fc
